### DISCUSSION TITLE
Integrating Python Matplotlib Heatmap into a JavaScript Application

### DISCUSSION THREAD

**Discussion by dolev146** [pyodide/pyodide#4494] [Q&A]
I am currently working with a medical company that is transitioning from MATLAB to JavaScript for our software solutions. As part of this process, we are looking to integrate heatmap visualizations similar to what we had in MATLAB.

like the heatmap shown in the attached image. I'm trying to figure out if there's a way to incorporate a Matplotlib heatmap into our new JavaScript application. Is there an approach or a tool that allows embedding or converting these Matplotlib heatmaps for use in a JavaScript environment?

Attached is an example of the heatmap we're working with, generated in Python using Matplotlib.

![Example of a Python Matplotlib Heatmap][1]

  [1]: https://i.stack.imgur.com/XQhrd.png

https://stackoverflow.com/questions/77960528/integrating-python-matplotlib-heatmap-into-a-javascript-application

**Comment by ryanking13** [comment]
Here is an example you can try. (For the Python part, I asked ChatGPT to write some heatmap code).

```html
<!doctype html>
<html>
  <head>
      <script src="https://cdn.jsdelivr.net/pyodide/v0.25.0/full/pyodide.js"></script>
  </head>
  <body>
    <script type="text/javascript">
      async function main(){
        let pyodide = await loadPyodide();
        await pyodide.loadPackage(["matplotlib", "numpy"]);
        const png = pyodide.runPython(`
import matplotlib.pyplot as plt
import numpy as np
from io import BytesIO
import base64

x = np.linspace(20, 40, 100)
y = np.linspace(0, 1, 100)
X, Y = np.meshgrid(x, y)
Z = np.exp(-((X-30)**2 + (Y-0.5)**2)/10)

# Create a heatmap plot
plt.figure(figsize=(8, 6))
c = plt.pcolormesh(X, Y, Z, cmap='jet', shading='auto')
plt.colorbar(c, label='z-dependent variable')
plt.xlabel('Separation distance')
plt.ylabel('(Z - Z_min) / (Z_max - Z_min)')

# Inverting the x-axis to match the example image
plt.gca().invert_xaxis()

# Show the plot
buf = BytesIO()
plt.savefig(buf, format='png')

base64.b64encode(buf.getvalue()).decode('utf-8')
        `)

        const img = document.createElement('img');
        img.src = 'data:image/png;base64,' + png;
        document.body.appendChild(img);
      }
      main();
    </script>
  </body>
</html>
```

Basically, the process of generating a plot is the same, the only difference is the process of displaying it on the screen.

```python
buf = BytesIO()
plt.savefig(buf, format='png')

base64.b64encode(buf.getvalue()).decode('utf-8')
```

It is worth noting that using `plt.show()`  in Pyodide will probably not work as you expect.

This is because we don't know where to plot: where on the webpage should the matplotlib add the element? So instead, I recommend you save the plot in an image file (bytes), and deal with it directly. In the example above, I encoded the bytes into `base64` string, which HTML img tag can accept. `pyodide.runPython` will return the base64-encoded plot (png image), and you can do whatever you want with it.
